src/crimson/os/seastore/omap_manager/btree/btree_omap_manager.cc

```diff
@@ -0,0 +1,172 @@
+      logger().debug("{}:  {} -> {}", __func__,  ret.first, ret.second);
+      return omap_set_key_ret(
+        omap_set_key_ertr::ready_future_marker{},
+        std::move(ret));
+  });
+
+}
+
+BtreeOMapManager::omap_rm_key_ret
+BtreeOMapManager::omap_rm_key(omap_root_t &omap_root, Transaction &t, const std::string &key)
+{
+  logger().debug("{}: {}", __func__, key);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, &key](auto extent) {
+    return extent->rm_key(get_omap_context(omap_root, t), key);
+  }).safe_then([](auto ret) {
+    logger().debug("{}: {}", __func__, ret);
+    return omap_rm_key_ret(
+      omap_rm_key_ertr::ready_future_marker{},
+      ret);
+  });
+
+}
+
+BtreeOMapManager::omap_list_keys_ret
+BtreeOMapManager::omap_list_keys(omap_root_t &omap_root, Transaction &t,
+                                 std::string start, size_t max_result_size)
+{
+  logger().debug("{}", __func__);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, start, max_result_size]
+    (auto extent) {
+    return extent->list_keys(get_omap_context(omap_root, t), start, max_result_size)
+      .safe_then([](auto &&result) {
+      return omap_list_keys_ret(
+             omap_list_keys_ertr::ready_future_marker{},
+             std::move(result));
+    });
+  });
+
+}
+
+BtreeOMapManager::omap_list_ret
+BtreeOMapManager::omap_list(omap_root_t &omap_root, Transaction &t,
+                            std::string start, size_t max_result_size)
+{
+  logger().debug("{}", __func__);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t, start, max_result_size]
+    (auto extent) {
+    return extent->list(get_omap_context(omap_root, t), start, max_result_size)
+      .safe_then([](auto &&result) {
+      return omap_list_ret(
+             omap_list_ertr::ready_future_marker{},
+             std::move(result));
+    });
+  });
+}
+
+BtreeOMapManager::omap_clear_ret
+BtreeOMapManager::omap_clear(omap_root_t &omap_root, Transaction &t)
+{
+  logger().debug("{}", __func__);
+  return get_omap_root(omap_root, t).safe_then([this, &omap_root, &t](auto extent) {
+    return extent->clear(get_omap_context(omap_root, t));
+  }).safe_then([this, &omap_root, &t] {
+    return tm.dec_ref(t, omap_root.omap_root_laddr).safe_then([&omap_root] (auto ret) {
+      omap_root.state = omap_root_state_t::MUTATED;
+      omap_root.depth = 0;
+      omap_root.omap_root_laddr = L_ADDR_NULL;
+      return omap_clear_ertr::now();
+    });
+  });
+}
+
+BtreeOMapManager::insert_key_ret
+BtreeOMapManager::insert_key(omap_root_t &omap_root, Transaction &t,
+                OMapNodeRef root, const std::string &key, const std::string &value)
+{
+  auto split = insert_key_ertr::make_ready_future<OMapNodeRef>(root);
+  bool overflow;
+  if (omap_root.depth >1)
+    overflow = root->extent_is_overflow(key.size() + 1);
+  else
+    overflow = root->extent_is_overflow(key.size() + 1 + value.size() + 1);
+  if (overflow) {
```

Comment:
I'm a bit worried about the split and merge behavior in general.  The LBA and extent trees both use a fairly simple strategy of splitting and merging during descent.  This works there because we know that the keys are fixed size and a small portion of the overall size.  Here, I think that if the keys and values are somewhat large compared with the size of the nodes, you'll end up with a lot of nodes significantly under the minimum size.

I think you actually noticed this problem partially in merge_entry with checking_parent().  With LBA and extent map trees, all keys are the same size, so changing the pivot cannot cause an overflow.  This is not necessarily the case with variably sized keys as you noted.  I think you need to:
- Remove the split and merge checks on the way to the leaf entirely
- Modify the inner node mutation methods to perform the mutation (insert, remove) and return a status indicating whether a split or merge needs to happen -- something like

```
struct mutation_result_t {
  emum result { SUCCESS, NEED_SPLIT, NEED_MERGE };

  /// Only populated if NEED_SPLIT, indicates the newly created left and right nodes
  /// from splitting the target entry during insertion.
  std::optional<std::pair<OmapInnerNodeRef, OmapInnerNodeRef>> split_nodes; 
};
```

I think this will substantially simplify your implementation.  The leaf that